### PR TITLE
add first base for reordering

### PR DESCRIPTION
**Pull request by petersg83**
### What does it do?

First base to allow partial update and reordering of relations

### CODE REVIEW

File: packages/core/database/lib/entity-manager.js [hunk 1]
```diff
@@ -615,40 +627,117 @@ const createEntityManager = (db) => {
...
-            // if there is nothing to insert
-            if (insert.length === 0) {
-              continue;
-            }
+              for (const relToDelete of existingRelsToDelete) {
+                await this.createQueryBuilder(joinTable.name)
+                  .update({ order: db.getConnection().raw('?? - 1', 'order') })
+                  .where({
+                    [joinColumn.name]: id,
+                    order: { $gt: relToDelete.order },
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager.js:671]: do we actually need an order column for this first iteration?
We could just have no order and add the new ones at the end.

File: packages/core/database/lib/entity-manager.js [hunk 2]
```diff
@@ -615,40 +627,117 @@ const createEntityManager = (db) => {
...
+                  .max('order')
+                  .where({ [joinColumn.name]: id })
+                  .where(joinTable.on || {})
+                  .first()
+                  .execute();
+
+                insert = _.uniqBy('id', connect).map((rel, idx) => ({
+                  [joinColumn.name]: id,
+                  [inverseJoinColumn.name]: rel.id,
+                  order: max + idx + 1,
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager.js:688]: Yes I think, otherwise how do you keep the order stable?

[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/database/lib/entity-manager.js:688]: If you are only adding new relations at the end , the order will always be stable I think

[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager.js:688]: It depends on the database actually, it tends to be sorted by creation date but there is no guarantee to it.
What we could do is to sort by id (since the id is incremental) that would work. However since we need to keep the order from the 2 sides, we need the 2 order columns and the id would not be enough then. So we almost could have avoided the order columns yes :/
(the 2 columns are not implemented yet)

[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/database/lib/entity-manager.js:688]: True, did not think about the order could be different in the database!

**Review by alexandrebodin — COMMENTED** [review]
Just a 1st pass for now. 

Can you add some more comments in your code to explain the ordering algorithm a bit more explicitly and add the jsdoc to the new functions you created.

We will have to refacto the code to move the realtional logic more in it's own file.

**Review by alexandrebodin — COMMENTED** [review]
Just a 1st pass for now. 

Can you add some more comments in your code to explain the ordering algorithm a bit more explicitly and add the jsdoc to the new functions you created.

We will have to refacto the code to move the realtional logic more in it's own file.

File: packages/core/content-manager/server/tests/index.test.e2e.js [hunk 3]
```diff
@@ -91,15 +91,11 @@ describe('Relations', () => {
       expect(body.id).toBeDefined();
       expect(body.name).toBe('tag1');
       expect(body.createdBy).toMatchObject({
-        firstname: 'admin',
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/content-manager/server/tests/index.test.e2e.js:94]: No, the firstname was incorrect. I believe it's because the test order changed and the firstname/lastname were changed in a previous test. So I just removed them as I don't think it's relevant to test that here.

File: packages/core/database/lib/entity-manager/index.js [hunk 4]
```diff
@@ -485,11 +504,38 @@ const createEntityManager = (db) => {
...
+            insert.forEach((rel, idx) => {
+              rel[orderColumnName] = idx + 1;
+            });
+          }
+          // add inv_order value when relevant
+          if (isBidirectional(attribute) && isManyToAny(attribute)) {
+            const maxMap = {};
+            await Promise.all(
+              relIdsToadd.map(async (relId) => {
+                const { max } = await this.createQueryBuilder(joinTable.name)
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:518]: yes totally, it's in my todo. However I think I need to lock the table, the transaction would not be enough. Knex doesn't allow table locking. I might need to do raw queries but I don't like it. Any idea of how to do differently?

File: packages/core/database/lib/entity-manager/index.js [hunk 5]
```diff
@@ -466,17 +481,21 @@ const createEntityManager = (db) => {
...
           const { joinTable } = attribute;
-          const { joinColumn, inverseJoinColumn } = joinTable;
-
-          if (isOneToAny(attribute) && isBidirectional(attribute)) {
-            await this.createQueryBuilder(joinTable.name)
-              .delete()
-              .where({ [inverseJoinColumn.name]: castArray(data[attributeName]) })
-              .where(joinTable.on || {})
-              .execute();
+          const { joinColumn, inverseJoinColumn, orderColumnName, inverseOrderColumnName } =
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:484]: You're talking about when there is no joinTable?

File: packages/core/database/lib/entity-manager/index.js [hunk 6]
```diff
@@ -644,42 +690,196 @@ const createEntityManager = (db) => {
...
+              let disconnect = uniqBy('id', cleanRelationData.disconnect || []);
+              if (isAnyToOne(attribute)) {
+                connect = connect.slice(-1);
+                disconnect = [];
+              }
+              relIdsToaddOrMove = toIds(connect);
+              // DELETE relations in disconnect
+              const relIdsToDelete = toIds(differenceBy('id', disconnect, connect));
+
+              await deleteAllRelations({ id, attribute, joinTable, onlyFor: relIdsToDelete, db });
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:724]: ahah yes I struggled a bit with this one :p
I will come up with something else

File: packages/core/database/lib/entity-manager/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/database/lib/entity-manager/index.js:518]: We might be able to use lock for update. The max is computed based on all rows so it might prevent reading it again until the transaction has ended. Not entirely sure though

File: packages/core/database/lib/entity-manager/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/database/lib/entity-manager/index.js:484]: yes. I see we won't be adding a order column next to it and that's fine but we should support the connect/disconnect syntax.

File: packages/core/database/lib/entity-manager/index.js [hunk 7]
```diff
@@ -644,42 +690,196 @@ const createEntityManager = (db) => {
...
+                      .execute();
+
+                    currentMovingRels.forEach((rel) => {
+                      if (rel[orderColumnName] > currentRel[orderColumnName]) {
+                        rel[orderColumnName] -= 1;
+                      }
+                    });
+                  }
+
+                  await this.createQueryBuilder(joinTable.name)
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:771]: Indeed, when `currentOrderIsNull` is true I should increment the max 👍

File: packages/core/database/lib/entity-manager/index.js [hunk 8]
```diff
@@ -644,42 +690,196 @@ const createEntityManager = (db) => {
...
+                .where(joinTable.on || {})
+                .execute();
+              const currentMovingRelsMap = groupBy(inverseJoinColumn.name, currentMovingRels);
+
+              let max = 0;
+              for (const relToAdd of relsToAdd) {
+                const currentRel = currentMovingRelsMap[relToAdd.id]?.[0];
+
+                if (currentRel && isAnyToMany(attribute)) {
+                  const update = { [orderColumnName]: max + 1 };
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:828]: yes it's the same here. max starts at `0`. I kept the same name for the variable to understand that the code is similar. I can call it index instead

File: packages/core/database/lib/entity-manager/index.js [hunk 9]
```diff
@@ -644,42 +690,196 @@ const createEntityManager = (db) => {
...
+                    ...(joinTable.on || {}),
+                    ...(relToAdd.__pivot || {}),
+                  };
+
+                  if (isAnyToMany(attribute)) {
+                    insert[orderColumnName] = max + 1;
+                  }
+                  // can be optimized in one query
+                  if (isBidirectional(attribute) && isManyToAny(attribute)) {
+                    const { max: reverseMax } = await this.createQueryBuilder(joinTable.name)
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:850]: Yes, thus the comment above. I have to think about it, see if the max won't changed in between. I hope it will work 👌

[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/database/lib/entity-manager/index.js:850]: Ho ok I missed the let max = 0 and the increment at the end of the loop. You could change the loop to use that idx to make it even more obvious we are talking about the relation index in the received data

File: packages/core/database/lib/entity-manager/index.js [hunk 10]
```diff
@@ -644,42 +690,196 @@ const createEntityManager = (db) => {
...
+                      .execute();
+
+                    insert[inverseOrderColumnName] = reverseMax + 1;
+                  }
+
+                  await this.createQueryBuilder(joinTable.name).insert(insert).execute();
+                  max += 1;
+                }
+              }
+            } else {
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:807]: yes I can explore that. It will require to have a composite foreign key on the join table

File: packages/core/database/lib/metadata/relations.js [hunk 11]
```diff
@@ -398,12 +401,20 @@ const createJoinTable = (metadata, { attributeName, attribute, meta }) => {
...
   let inverseJoinColumnName = _.snakeCase(`${targetMeta.singularName}_id`);
 
-  // if relation is slef referencing
+  // if relation is self referencing
   if (joinColumnName === inverseJoinColumnName) {
     inverseJoinColumnName = `inv_${inverseJoinColumnName}`;
   }
 
-  metadata.add({
+  const orderColumnName = _.snakeCase(`${meta.singularName}_order`);
```
Comment (alexandrebodin) [inline packages/core/database/lib/metadata/relations.js:409]: Should this be the opposite naming? if Articles have categories shouldn't this be categories_order here, and the inverse be articles_categories?

File: packages/core/database/lib/metadata/relations.js [hunk 12]
```diff
@@ -398,12 +401,20 @@ const createJoinTable = (metadata, { attributeName, attribute, meta }) => {
...
+  let inverseOrderColumnName = _.snakeCase(`${targetMeta.singularName}_order`);
+
+  // if relation is self referencing
+  if (attribute.relation === 'manyToMany' && joinColumnName === inverseJoinColumnName) {
```
Comment (alexandrebodin) [inline packages/core/database/lib/metadata/relations.js:413]: Why are you only testing on manyToMany ?

File: packages/core/database/lib/entity-manager/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:532]: Interesting, it's deadlocking here even though not on the same table or row.
It might be just sqlite however, I need to test with postgres and mysql.
Capture d’écran 2022-09-19 à 15 55 33

[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:532]: Indeed with postgres it's not the same. More than 1 connection I believe
Capture d’écran 2022-09-19 à 16 07 12

[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:532]: `forUpdate` doesn't prevent from reading the line :/ 

Capture d’écran 2022-09-19 à 16 20 14

File: packages/core/database/lib/metadata/relations.js (hunk 11 above)
Comment (petersg83) [reply packages/core/database/lib/metadata/relations.js:409]: I think both are possible as we can also see it as : what is the order from the article side?
I think I prefer to inverse however, it's move intuitive

File: packages/core/database/lib/entity-manager/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:537]: https://stackoverflow.com/questions/52325771/manually-lock-table-in-knex

Capture d’écran 2022-09-19 à 19 00 57

Mmmh.. It won't be solved easily..

File: packages/core/database/lib/entity-manager/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js]: What do you think of it now?

File: packages/core/database/lib/metadata/index.js [hunk 13]
```diff
@@ -142,6 +142,11 @@ const createCompoLinkModelMeta = (baseModelMeta) => {
         name: `${baseModelMeta.tableName}_entity_fk`,
         columns: ['entity_id'],
       },
+      {
+        name: `${baseModelMeta.tableName}_unique`,
+        columns: ['entity_id', 'component_id', 'field'],
+        type: 'unique',
+      },
```
Comment (Marc-Roig) [inline packages/core/database/lib/metadata/index.js:149]: just curious, is this index related to relations? 
I am trying to understand if it is related to the reordering.

Comment (petersg83) [reply packages/core/database/lib/metadata/index.js]: It's related to relations yes! It's for 2 things:
- making sure we don't have twice the same relation has it can lead to unexpected results
- be able to use the SQL/knex `onConflict` method: we insert all rows and if it's already existing then we just update the order. It can help for shorter code and maybe performance

File: packages/core/database/lib/metadata/relations.js [hunk 14]
```diff
@@ -433,6 +444,11 @@ const createJoinTable = (metadata, { attributeName, attribute, meta }) => {
         name: `${joinTableName}_inv_fk`,
         columns: [inverseJoinColumnName],
       },
+      {
+        name: `${joinTableName}_unique`,
+        columns: [joinColumnName, inverseJoinColumnName],
+        type: 'unique',
+      },
     ],
```
Comment (Marc-Roig) [inline packages/core/database/lib/metadata/relations.js:452]: This is more of a question regarding metadata.
What I understood about it was that it is an object that represents the schema in the database + other info.

Are you creating an index by adding this?

Comment (petersg83) [reply packages/core/database/lib/metadata/relations.js:452]: yes exactly :) it's what is used to create the database tables

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+        .limit(batchSize)
+        .execute();
+      done = relsToDelete.length < batchSize;
+      lastId = relsToDelete[relsToDelete.length - 1]?.id;
+
+      // ORDER UPDATE
+      if (isAnyToMany(attribute)) {
+        // sort by order DESC so that the order updates are done in the correct order
+        // avoiding one to interfere with the others
+        relsToDelete.sort((a, b) => b[orderColumnName] - a[orderColumnName]);
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/regular-relations.js:145]: why not orderingBy `ordingColumnName` in the query above?

File: packages/core/database/lib/entity-manager/index.js [hunk 16]
```diff
@@ -458,25 +495,28 @@ const createEntityManager = (db) => {
...
-              .delete()
-              .where({ [inverseJoinColumn.name]: castArray(data[attributeName]) })
-              .where(joinTable.on || {})
-              .execute();
+          const { joinColumn, inverseJoinColumn, orderColumnName, inverseOrderColumnName } =
+            joinTable;
+          const select = [joinColumn.name];
+          if (isAnyToMany(attribute)) {
+            select.push(orderColumnName);
           }
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/index.js:511]: I hate to be this type of person. But I am missing some comments here giving context 😬
I understand you are selecting the order column only if the relationship is between many others? So then you increment that order.

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+            .decrement(orderColumnName, 1)
+            .where({
+              [joinColumn.name]: relToDelete[joinColumn.name],
+              [orderColumnName]: { $gt: relToDelete[orderColumnName] },
+            })
+            .where(joinTable.on || {})
+            .execute();
+        }
+      }
+    }
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/regular-relations.js:55]: Maybe it makes this too complicated to read, but I think this can be done with a single query.
https://github.com/strapi/experiments/pull/1/files#diff-c253cf0e2025afa2af251a04e4dcbe8deeec721161e95ef719b758bc2cc78e90R24

Didn't spend time doing it with knexjs query builder but I can try if you find it addecuate.

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+      const currentRelsToDelete = await createQueryBuilder(joinTable.name, db)
+        .select(select)
+        .where({
+          [inverseJoinColumn.name]: relIdsToadd,
+          [joinColumn.name]: { $ne: id },
+        })
+        .where(joinTable.on || {})
+        .execute();
+
+      currentRelsToDelete.sort((a, b) => b[orderColumnName] - a[orderColumnName]);
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/regular-relations.js:41]: again, same question as before, why not ordering it in the query above?

File: packages/core/database/lib/entity-manager/index.js [hunk 19]
```diff
@@ -592,7 +661,7 @@ const createEntityManager = (db) => {
             })
             .execute();
 
-          const rows = toAssocs(data[attributeName]).map((data) => ({
+          const rows = cleanRelationData.set.map((data) => ({
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/index.js:664]: I think you should check if cleanRelationData is empty.
It is giving me an error if it is (trying to update a relation from a kitchensink type in getstarted)

Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:664]: Indeed 👌  I'll fix that when adding connect/disconnect support to morph relations

File: packages/core/database/lib/entity-manager/regular-relations.js (hunk 18 above)
Comment (petersg83) [reply packages/core/database/lib/entity-manager/regular-relations.js:41]: good idea 👌

File: packages/core/database/lib/entity-manager/regular-relations.js (hunk 17 above)
Comment (petersg83) [reply packages/core/database/lib/entity-manager/regular-relations.js:55]: Wow, thank you very much, I'll use that! 😀

Comment (Marc-Roig) [reply packages/core/database/lib/entity-manager/regular-relations.js:55]: Wow, row_number rocks. Forgot about that one.
Makes complete sense. You just reorder everything from the start with a single query, no decrements, no math.

Comment (Marc-Roig) [reply packages/core/database/lib/entity-manager/regular-relations.js:55]: Actually, you do not need the partition by, only with:
```
SELECT id, ROW_NUMBER() OVER (ORDER BY order) AS rowNum FROM tes_table
```

Comment (alexandrebodin) [reply packages/core/database/lib/entity-manager/regular-relations.js:55]: I think we need the partition here 🤔 it won't give the right order per document but a global order for all the relations of all documents otherwise.

Comment (Marc-Roig) [reply packages/core/database/lib/entity-manager/regular-relations.js:55]: True, forgot we had two order columns

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 20]
```diff
@@ -0,0 +1,197 @@
+'use strict';
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/regular-relations.js:1]: There is another file called "morph-relations", so I didn't know how to call this one :P it's similar, just that it handles the "regular" relations :p

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+          [joinColumn.name]: { $ne: id },
+        })
+        .where(joinTable.on || {})
+        .execute();
+
+      currentRelsToDelete.sort((a, b) => b[orderColumnName] - a[orderColumnName]);
+
+      for (const relToDelete of currentRelsToDelete) {
+        if (relToDelete[orderColumnName] !== null) {
+          await createQueryBuilder(joinTable.name, db)
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/regular-relations.js:45]: I asked if we want to be less strict. We didn't conclude so I went for the strict way as we can easily go the loose way if needed later. If you believe we should do loose right away we can discuss that

[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/database/lib/entity-manager/regular-relations.js:45]: I think with the proposals below we will be fine all good for me :)

File: packages/core/database/lib/entity-manager/regular-relations.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/database/lib/entity-manager/regular-relations.js:1]: Ho ok I missed the other name. We probably will split them further later on that's fine by me

File: packages/core/database/lib/entity-manager/index.js [hunk 22]
```diff
@@ -633,53 +702,242 @@ const createEntityManager = (db) => {
...
+                .max(inverseOrderColumnName, { as: 'max' })
+                .whereIn(inverseJoinColumn.name, nonExistingRelsIds)
+                .where(joinTable.on || {})
+                .groupBy(inverseJoinColumn.name)
+                .from(joinTable.name);
+
+              const maxMap = maxResults.reduce(
+                (acc, res) => Object.assign(acc, { [res[inverseJoinColumn.name]]: res.max }),
+                {}
+              );
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/index.js:788]: I think here we should only execute this if join table has the inverse order column
```suggestion
            let maxMap = {}
            if ( isAnyToMany(attribute) && isBidirectional(attribute)) {  
              const maxResults = await db
                .getConnection()
                .select(inverseJoinColumn.name)
                .max(inverseOrderColumnName, { as: 'max' })
                .whereIn(inverseJoinColumn.name, nonExistingRelsIds)
                .where(joinTable.on || {})
                .groupBy(inverseJoinColumn.name)
                .from(joinTable.name);

              maxMap = maxResults.reduce(
                (acc, res) => Object.assign(acc, { [res[inverseJoinColumn.name]]: res.max }),
                {}
              );
         }
```

Comment (Marc-Roig) [reply packages/core/database/lib/entity-manager/index.js:788]: Also, I think it would improve the readability if we change:
`isManyToAny(attribute) && isBidirectional(attribute)` -> `hasInverseOrderColumn()`
`isAnyToMany(attribute)`-> `hasOrderColumn()`

File: packages/core/database/lib/entity-manager/index.js [hunk 23]
```diff
@@ -364,17 +397,19 @@ const createEntityManager = (db) => {
             // set columns
             const { idColumn, typeColumn } = targetAttribute.morphColumn;
 
+            const relId = toId(cleanRelationData.set[0]);
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/index.js:400]: We should check if cleanRelationData is null here.

File: packages/core/database/lib/entity-manager/index.js [hunk 24]
```diff
@@ -429,13 +465,14 @@ const createEntityManager = (db) => {
         }
 
         if (attribute.joinColumn && attribute.owner) {
+          const relIdsToAdd = toIds(cleanRelationData.set);
           if (
             attribute.relation === 'oneToOne' &&
             isBidirectional(attribute) &&
-            data[attributeName]
+            relIdsToAdd.length
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/index.js:472]: This is a question maybe for @alexandrebodin . why we only add the relation when its a `oneToOne` when using `joinColumn`

File: packages/core/database/lib/entity-manager/index.js [hunk 25]
```diff
@@ -633,53 +700,245 @@ const createEntityManager = (db) => {
...
+          if (isNull(cleanRelationData.set)) {
+            await deleteRelations({ id, attribute, joinTable, db }, { relIdsToDelete: 'all' });
+          } else {
+            const isPartialUpdate = !has('set', cleanRelationData);
+            let relIdsToaddOrMove;
+
+            if (isPartialUpdate) {
+              if (isAnyToOne(attribute)) {
+                cleanRelationData.connect = cleanRelationData.connect.slice(-1);
+                cleanRelationData.disconnect = [];
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/index.js:734]: 📝 Note from the meeting we had
Disconnect should not be empty here, for AnyToOne relations you can not remove the relationship and leave it empty.

File: packages/core/database/lib/entity-manager/index.js (hunk 24 above)
Comment (alexandrebodin) [reply packages/core/database/lib/entity-manager/index.js:472]: We actually add it for oneToOne or oneToMany (look a bit below in the code :)) 

This usecase is just to unset the relations on oneToOne because it must be maintain as a unique on both side contrary to oneToMany where it will be an additive operation

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+  }
+  return select;
+};
+
+const deletePreviousOneToAnyRelations = async ({ id, attribute, joinTable, relIdsToadd, db }) => {
+  const { joinColumn, inverseJoinColumn, orderColumnName } = joinTable;
+  const select = getSelect(joinTable, attribute);
+
+  // need to delete the previous relations for oneToAny relations
+  if (isBidirectional(attribute) && isOneToAny(attribute)) {
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/regular-relations.js]: done 👌

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+    // update orders for previous oneToAny relations that will be deleted if it has order (oneToMany)
+    if (isAnyToMany(attribute)) {
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/database/lib/entity-manager/regular-relations.js]: I would say yes and no.
- `anyToMany`: don't have a good idea of context as we might think it's possible to have `oneToOne`. However we know we are supposed to end up in this if anyway.
- `oneToMany`: don't understand exactly why the `oneToMany` is treated but not the other relations. But we better understand what happens in the if.

In both situations, to understand what happens, we need to know that we are in a `oneToAny`. So I would say it's equivalent. I hoped the comment line 30 would be enough because I gave the information that we are in `oneToMany`.

File: packages/core/database/lib/entity-manager/index.js (hunk 23 above)
Comment (petersg83) [reply packages/core/database/lib/entity-manager/index.js:407]: done :)

**Comment by petersg83** [comment]
There are still this things to do:
  - Add connect/disconnect support to morphRelations
  - Transaction
  - improve comments
  - Finish simplifying the code
  
 It will be done in another PR later

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 28]
```diff
@@ -0,0 +1,257 @@
...
+    update.push('?? = b.inv_order');
+    updateBinding.push(inverseOrderColumnName);
+    select.push('ROW_NUMBER() OVER (PARTITION BY ?? ORDER BY ??) AS inv_order');
+    selectBinding.push(inverseJoinColumn.name, inverseOrderColumnName);
+    where.push(`?? IN (${inverseRelIds.map(() => '?').join(', ')})`);
+    whereBinding.push(inverseJoinColumn.name, ...inverseRelIds);
+  }
+
+  // raw query as knex doesn't allow updating from a subquery
+  // https://github.com/knex/knex/issues/2504
```
Comment (Marc-Roig) [inline packages/core/database/lib/entity-manager/regular-relations.js:208]: I think it is not applicable to this query. But with Alex we managed to do an update like this with knex (for the migrations)

````
    const subQuery = knex
      .select("id")
      .rowNumber("order_A", "id", "tableA_id")
      .from("links")
      .as("partition");

    await knex
      .update(
        "order_A",
        knex
          .select("order_A")
          .from(subQuery)
          .where(knex.ref("links.id"), knex.ref("partition.id"))
      )
      .from("links");
      ```

**Comment by Marc-Roig** [comment]
I think the order in this case is not preserved:
image
- I have `Tags` in a `Kitchen Sink` entry. T1, T2, T3.
- I remove T2
- I save
- I add again T2 (placed in the end in the UI)
- I save
- I see T2 in the middle again.

**Comment by Marc-Roig** [comment]
nice! LGTM then :)

**Review by Marc-Roig — APPROVED** [review]
